README.md
jslogo - Logo in JavaScript
===========================

This is hosted at http://calormen.com/jslogo for playing with live.

Logo Examples
-------------
    to star repeat 5 [ fd 100 rt 144 ] end
    star
    to square :length repeat 4 [ fd :length rt 90 ] end
    repeat 36 [ square 50 rt 10 ]
    to randomcolor setcolor pick [ red orange yellow green blue violet ] end
    repeat 36 [ randomcolor square random 200 rt 10 ]
    window pu repeat 72 [ setlabelheight repcount fd repcount * 2.5 label "Logo bk repcount * 2.5 rt 10 ]





ChangeLog
== Change Log ==

NOTE: Changes after December 2011 are tracked in the jslogo version control change history.

2011-12-29
	Added support for IE8 via FlashCanvas
	Migrated unit tests over to use QUnit.

2011-10-22
	Added another feature from Ofer Wald: ARC angle radius (example)

2011-10-14
	Added COPYDEF and made it and ERASE respect REDEFP when it comes to primitives (built-in procedures).
	Applied two patches from Ofer Wald: make text drawn with LABEL (example) rotate with the mouse, and prevent primitives from being overridden with TO
	You can now put short programs in the URL hash (#blah) as well as the query string (?blah).

2010-08-07
	Modified the way the turtle is drawn - it now uses an overlay canvas instead of drawing to a backbuffer. This simplifies the code. (I didn't realize HTML5 canvases were transparent when I first implemented this!)
	Removed fallback canvas text API support, so LABEL will not work on iOS 3 and earlier or Firefox 3.0.

2010-03-27
	Added: WINDOW, WRAP, FENCE and TURTLEMODE

2009-11-26
	Infix arithmetic and order-comparison operators now force arguments to be numbers.
	Fractional decimals with no leading zero (e.g. .1) now parse correctly again.
	Added: PENPAINT, PENERASE, PENREVERSE
	READWORD can take an optional prompt string input. e.g. (READWORD "width:)

2009-11-24
	Expression evaluation separated from parsing. AND and OR now short-circuit evaluation.
	Slight tweaks to TYPE output.
	Added aliases: BF, BFS, BL, IFT, IFF, OP
	Added: READWORD, ABS

2009-11-19
	You can now specify a program to run in the URL hash or query string. You can %-encode characters or, as a shortcut, use _ instead of space, e.g. http://www.calormen.com/Logo/?repeat_4_[fd_100_rt_90]
	Thanks to discussion on LogoForum has shown me the error of my ways regarding lists. The contents of list literals is no longer evaluated.

2009-11-11
	UCBLogo alignment - almost done!
	Added: run, runresult, test, iftrue, iffalse, .maybeoutput, ignore, for, do.while, while, do.until, until

2009-11-10
	UCBLogo alignment continues....
	Added: contents, procedures, primitives, globals, names, erase, erall, apply, invoke, foreach, find
	Also, fixed a bug with word on WebKit, and ensured that long text output wraps.

2009-11-09
	UCBLogo alignment continies....
	Added: name, local, localmake, thing, global, procedurep, primitivep, definedp, namep, forever, bye
	Corrected stop - was like bye, now like an empty output

2009-11-08
	UCBLogo alignment continies....
	Added: remainder, modulo, exp, log10, ln, radarctan, radsin, radcos, radtan, rseq, bitand, bitor, bitxor, bitnot, ashift, lshift, label, setlabelheight, shownp, labelsize, pendownp, pencolor, pensize
	Removed: modulus (use remainder or modulo), write (use label), setfont (use setlabelheight)
	Corrected argument order for (arctan x y)
	In (setpencolor r g b), r/g/b must be in 0..99; Logo colors 0...15 are supported.

2009-11-07
	I've started to work on aligning the syntax with UCB Logo. Language reference now follows UCB Logo documentation structure for ease of comparison. Code and unit tests similarly reorganized.
	Added: sentence, combine, gensym, firsts, butfirsts, item, remove, remdup, wordp, listp, numberp, count, ascii, char, lowercase, uppercase
	Added aliases setpensize, setpencolor
	Fixed equality comparisons
	Added: push, pop, queue, dequeue, emptyp, beforep, memberp, substringp
	Rigged up a text screen overlay.
	Added: print, type, show, cleartext, emptyp, beforep, memberp, substringp
	Use write for turtle graphics text now

2009-10-31
	A series of appropriately spooky changes - they're there, but you can't see them!
	Change log is now accessible as an Atom feed
	Switch to native Javascript types for primitive types (number, string, list)
	Cleaned up use of 'undefined', stopping
	Turtle graphics puled into own script module
	Pulled alerting on error out of interpreter
	Unit tests for unary minus, true, false
	Unary minus handling revamped - now confined to parser/serializer

2009-10-28
	def now takes a string argument for procname
	def now serializes operators correctly
	Fixed clean (was alerting)
	Fixed heading (was returning incorrect values)
	More unit tests

2009-10-27
	Found and fixed a handful of bugs while implementing unit tests:
	Operators <=, >=, <> were incorrectly parsed
	if, ifelse and repeat return now result of last evaluation (language extension?)
	and, or, xor now return 0 or 1
	lessequalp and lessequal? corrected

2009-10-26
	Fixed hiding of the language reference pane
	Corrected documentation - it's setfont not font
	Added text rendering for non-Mozilla browsers c/o Jim Studt's Hershey Font implementation for HTML5 canvas

2009-10-24
	Now licensed under the Apache 2.0 license

2009-04-13
	Fixed towards and setpos, which have been broken since the 2008-08-30 update (Sigh - I really need to implement unit tests.)

2009-03-21
	Unary minus disambiguation (similar to UCBLogo):

	Minus sign is parsed as unary minus if it begins a top-level expression, or if it follows an infix operator, or it has a leading space and has no trailing space.
	Otherwise, minus sign is parsed as an infix difference operator.
	Compare:
	cs ht print -5 (unary; will print -5)
	cs ht print - 5 + 10 (unary; will print 5)
	cs ht print 10 + - 5 (unary; will print 5)
	cs ht print 10 -5 (unary; will print 10)
	cs ht print 10 - 5 (infix; will print 5)
	cs ht print 10-5 (infix; will print 10)
	cs ht print 10- 5 (infix; will print 5)
	cs ht print sum 10 -5 (unary; will print 5)
	cs ht print sum 10 - 5 (infix; will produce error!)

2009-01-20
	Added list, fput, lput
	Added show (temporary) which returns a string representation of its argument

2008-09-28
	Added (temporary) display of command history, at the bottom of the window
	Added def (hacky implementation) that shows the definition of a function

2008-09-02
	Added reverse, iseq
	print can now print lists

turtle.js
//
// Turtle Graphics in Javascript
//

// Copyright (C) 2011 Joshua Bell
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function CanvasTurtle(canvas_ctx, turtle_ctx, width, height) {
  width = Number(width);
  height = Number(height);

  function deg2rad(d) { return d / 180 * Math.PI; }
  function rad2deg(r) { return r * 180 / Math.PI; }

  var self = this;
  function moveto(x, y) {

    function _go(x1, y1, x2, y2) {
      if (self.down) {
        canvas_ctx.beginPath();
        canvas_ctx.moveTo(x1, y1);
        canvas_ctx.lineTo(x2, y2);
        canvas_ctx.stroke();
      }
    }

    var ix, iy, wx, wy, fx, fy, less;

    while (true) {
      // TODO: What happens if we switch modes and turtle is outside bounds?

      switch (self.turtlemode) {
        case 'window':
          _go(self.x, self.y, x, y);
          self.x = x;
          self.y = y;
          return;

        default:
        case 'wrap':
        case 'fence':

          // fraction before intersecting
          fx = 1;
          fy = 1;

          if (x < 0) {
            fx = (self.x - 0) / (self.x - x);
          } else if (x >= width) {
            fx = (self.x - width) / (self.x - x);
          }

          if (y < 0) {
            fy = (self.y - 0) / (self.y - y);
          } else if (y >= height) {
            fy = (self.y - height) / (self.y - y);
          }

          // intersection point (draw current to here)
          ix = x;
          iy = y;

          // endpoint after wrapping (next "here")
          wx = x;
          wy = y;

          if (fx < 1 && fx <= fy) {
            less = (x < 0);
            ix = less ? 0 : width;
            iy = self.y - fx * (self.y - y);
            x += less ? width : -width;
            wx = less ? width : 0;
            wy = iy;
          } else if (fy < 1 && fy <= fx) {
            less = (y < 0);
            ix = self.x - fy * (self.x - x);
            iy = less ? 0 : height;
            y += less ? height : -height;
            wx = ix;
            wy = less ? height : 0;
          }

          _go(self.x, self.y, ix, iy);

          if (self.turtlemode === 'fence') {
            // FENCE - stop on collision
            self.x = ix;
            self.y = iy;
            return;
          } else {
            // WRAP - keep going
            self.x = wx;
            self.y = wy;
            if (fx === 1 && fy === 1) {
              return;
            }
          }

          break;
      }
    }
  }

  this.move = function(distance) {
    var x, y, point, saved_x, saved_y, EPSILON = 1e-3;

    point = Math.abs(distance) < EPSILON;

    if (point) {
      saved_x = this.x;
      saved_y = this.y;
      distance = EPSILON;
    }

    x = this.x + distance * Math.cos(this.r);
    y = this.y - distance * Math.sin(this.r);
    moveto(x, y);

    if (point) {
      this.x = saved_x;
      this.y = saved_y;
    }
  };

  this.turn = function(angle) {
    this.r -= deg2rad(angle);
  };

  this.penup = function() { this.down = false; };
  this.pendown = function() { this.down = true; };

  this.setpenmode = function(penmode) {
    this.penmode = penmode;
    canvas_ctx.globalCompositeOperation =
                (this.penmode === 'erase') ? 'destination-out' :
                (this.penmode === 'reverse') ? 'xor' : 'source-over';
  };
  this.getpenmode = function() { return this.penmode; };

  this.setturtlemode = function(turtlemode) { this.turtlemode = turtlemode; };
  this.getturtlemode = function() { return this.turtlemode; };

  this.ispendown = function() { return this.down; };

  var STANDARD_COLORS = {
    0: "black", 1: "blue", 2: "lime", 3: "cyan",
    4: "red", 5: "magenta", 6: "yellow", 7: "white",
    8: "brown", 9: "tan", 10: "green", 11: "aquamarine",
    12: "salmon", 13: "purple", 14: "orange", 15: "gray"
  };

  this.setcolor = function(color) {
    if (STANDARD_COLORS[color] !== undefined) {
      this.color = STANDARD_COLORS[color];
    } else {
      this.color = color;
    }
    canvas_ctx.strokeStyle = this.color;
    canvas_ctx.fillStyle = this.color;
  };
  this.getcolor = function() { return this.color; };

  this.setwidth = function(width) {
    this.width = width;
    canvas_ctx.lineWidth = this.width;
  };
  this.getwidth = function() { return this.width; };

  this.setfontsize = function(size) {
    this.fontsize = size;
    canvas_ctx.font = this.fontsize + 'px sans-serif';
  };
  this.getfontsize = function() { return this.fontsize; };

  this.setposition = function(x, y) {
    x = (x === undefined) ? this.x : x + (width / 2);
    y = (y === undefined) ? this.y : -y + (height / 2);

    moveto(x, y);
  };

  this.towards = function(x, y) {
    x = x + (width / 2);
    y = -y + (height / 2);

    return 90 - rad2deg(Math.atan2(this.y - y, x - this.x));
  };

  this.setheading = function(angle) {
    this.r = deg2rad(90 - angle);
  };

  this.clearscreen = function() {
    this.home();
    this.clear();
  };

  this.clear = function() {
    canvas_ctx.clearRect(0, 0, width, height);
  };

  this.home = function() {
    moveto(width / 2, height / 2);
    this.r = deg2rad(90);
  };

  this.showturtle = function() {
    this.visible = true;
  };

  this.hideturtle = function() {
    this.visible = false;
  };

  this.isturtlevisible = function() {
    return this.visible;
  };

  this.getheading = function() {
    return 90 - rad2deg(this.r);
  };

  this.getxy = function() {
    return [this.x - (width / 2), -this.y + (height / 2)];
  };

  this.drawtext = function(text) {
    canvas_ctx.save();
    canvas_ctx.translate(this.x, this.y);
    canvas_ctx.rotate(-this.r);
    canvas_ctx.fillText(text, 0, 0);
    canvas_ctx.restore();
  };

  this.arc = function(angle, radius) {
    var self = this;
    if (this.turtlemode == 'wrap') {
      [self.x, self.x + width, this.x - width].forEach(function(x) {
        [self.y, self.y + height, this.y - height].forEach(function(y) {
          canvas_ctx.beginPath();
          canvas_ctx.arc(x, y, radius, -self.r, -self.r + deg2rad(angle), false);
          canvas_ctx.stroke();
        });
      });
    } else {
      canvas_ctx.beginPath();
      canvas_ctx.arc(this.x, this.y, radius, -this.r, -this.r + deg2rad(angle), false);
      canvas_ctx.stroke();
    }
  };

  this.begin = function() {
    // Erase turtle
    turtle_ctx.clearRect(0, 0, width, height);

    // Stub for old browsers w/ canvas but no text functions
    canvas_ctx.fillText = canvas_ctx.fillText || function fillText(string, x, y) { };
  };

  this.end = function() {
    if (this.visible) {
      var ctx = turtle_ctx;
      ctx.beginPath();
      ctx.moveTo(this.x + Math.cos(this.r) * 20, this.y - Math.sin(this.r) * 20);
      ctx.lineTo(this.x + Math.cos(this.r - Math.PI * 2 / 3) * 10, this.y - Math.sin(this.r - Math.PI * 2 / 3) * 10);
      ctx.lineTo(this.x + Math.cos(this.r + Math.PI * 2 / 3) * 10, this.y - Math.sin(this.r + Math.PI * 2 / 3) * 10);
      ctx.lineTo(this.x + Math.cos(this.r) * 20, this.y - Math.sin(this.r) * 20);
      ctx.stroke();
    }
  };


  canvas_ctx.lineCap = 'round';

  turtle_ctx.lineCap = 'round';
  turtle_ctx.strokeStyle = 'green';
  turtle_ctx.lineWidth = 2;

  this.setcolor('#000000');
  this.setwidth(1);
  this.setpenmode('paint');
  this.setfontsize(14);
  this.setturtlemode('wrap');
  this.showturtle(true);
  this.pendown(true);

  this.x = width / 2;
  this.y = height / 2;
  this.r = Math.PI / 2;

  this.begin();
  this.end();
}